### ISSUE TITLE
SC2045 - bad advice?

### ISSUE THREAD

**Issue by matthewpersicowork** [koalaman/shellcheck#2154]
#### For bugs
- Rule Id (if any, e.g. SC1000): 
- My shellcheck version (`shellcheck --version` or 'online'): online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### For new checks and feature suggestions
- [x] shellcheck.net (i.e. the latest commit) currently gives no useful warnings about this
- [ ] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

THEDIR=/tmp/o/b/b

for i in "$THEDIR"/*
do
    echo "$i"
done
    
```

#### Here's what shellcheck currently says:
```

```
#### Here's what I wanted or expected to see:

I got to this point because shellcheck threw an SC2045 at me (Iterating over ls output is fragile. Use globs.) when I was using
```
for i in $(ls "$THEDIR"/*)
...
```
and told me to use a glob. However a glob does not work on an empty dir. To wit:
```
$ THEDIR=o/b/b

$ ls -la $THEDIR
total 0
drwxrwxr-x 2 mpersico5 users 17 Feb 19 14:44 .
drwxrwxr-x 3 mpersico5 users 15 Feb 19 14:36 ..
-rw-rw-r-- 1 mpersico5 users  0 Feb 19 14:44 foo

$ for i in "$THEDIR"/*; do     echo "$i"; done
o/b/b/foo

$ rm o/b/b/foo
rm: remove regular empty file 'o/b/b/foo'? y

$ for i in "$THEDIR"/*; do echo "$i"; done
o/b/b/*
```
I want the last result to be empty, not an unresolved glob string. I think the glob advice is potentially dangerous.

**Comment by matthewpersicowork** [comment]
And it's not the variable causing the issue:
```
$ for i in o/b/b/*; do     echo "$i"; done
o/b/b/bar
o/b/b/foo
$ for i in o/b/b/*; do    rm -f "$i"; done
$ for i in o/b/b/*; do     echo "$i"; done
o/b/b/*
```

**Comment by TinCanTech** [comment]
The same is true for `#!/bin/sh`

There is an obvious work-around but you appear to be questioning either shell behaviour or shellcheck behaviour for the specific case of an empty dir ?

**Comment by matthewpersico** [comment]
> Edit: Is this more to do with `glob` behaviour ?
It's not 'more' to do, it's about understanding  a subtlety, one which I don't believe I've encountered before myself in all the years I have been programming.
 
> Edit 2: In fact SC2045 documents the exact work-around ;-)
And it is misleading. To wit:
```
matthew@monolith:~/tmp$ touch a b c
matthew@monolith:~/tmp$ for i in *
> do echo $i;done
a
b
c
matthew@monolith:~/tmp$ rm a b c
matthew@monolith:~/tmp$ for i in *; do echo $i;done
*
matthew@monolith:~/tmp$
```
It is awfully misleading to have no files satisfy the glob, yet the loop runs once on the literal glob text. In my case, I was writing a debian helper, trying to link files and I ended up with a softlink to nowhere. Using a glob alone won't cut it.

**Comment by matthewpersico** [comment]
Update: My last sentence was prophetic. I started reading the bash manpage searching for 'glob' and I found this:
```
$ shopt nullglob
nullglob        off
$ for i in *; do echo $i;done
*
$ shopt -s nullglob
$ shopt nullglob
nullglob        on
$ for i in *; do echo $i;done
$
```

**Comment by matthewpersico** [comment]
Which is plastered right there on the 2045 page:
```
Also note that in Bash, shopt -s nullglob will allow the loop to run 0 times instead of 1 if there are no matches. There are also several other conditions to be aware of.
```
🤦

**Comment by TinCanTech** [comment]
That was the work-around I mentioned ;-)

**Comment by brother** [comment]
> Please consider the issue closed.

I think that you the reporter can close it.
